northstar/src/runtime/process/init.rs

```diff
@@ -30,11 +38,130 @@ pub(super) struct Init {
     pub seccomp: Option<AllowList>,
 }
 
+/// This is the type used to communicate with the init process from the runtime.
+#[derive(Debug, Serialize, Deserialize)]
+pub enum InitMessage {
+    Forked { pid: i32 },
+    Exit { pid: i32, exit_status: ExitStatus },
+}
+
+/// Init commands
+#[derive(Debug, Serialize, Deserialize)]
+pub enum InitApi {
+    /// Executes an command
+    Exec {
+        path: CString,
+        args: Vec<CString>,
+        env: Vec<CString>,
+    },
+}
+
+/// Init process main function
+fn init_main(
+    mut rt_tx: channel::Sender<InitMessage>,
+    mut rt_rx: channel::Receiver<InitApi>,
+    seccomp: Option<AllowList>,
+) -> ! {
+    enum Event {
+        Exit { pid: i32, exit_status: ExitStatus },
+        Command(InitApi),
+    }
+    let (tx, rx) = mpsc::channel();
+
+    // This thread forwards the commands coming from the runtime
+    let command_tx = tx.clone();
+    std::thread::spawn(move || loop {
+        let command = rt_rx
+            .recv()
+            .expect("Failed to receive init command from runtime");
+        command_tx.send(Event::Command(command)).unwrap();
+    });
+
+    // This thread sends the exit status of the child processes
+    let exit_status_tx = tx;
+    let reaper_thread = std::thread::spawn(move || {
+        loop {
+            match waitpid(None, None) {
+                Ok(WaitStatus::Exited(pid, status)) => {
+                    exit_status_tx
+                        .send(Event::Exit {
+                            pid: pid.as_raw(),
+                            exit_status: ExitStatus::Exit(status),
+                        })
+                        .unwrap();
+                }
+                Ok(WaitStatus::Signaled(pid, status, _)) => {
+                    exit_status_tx
+                        .send(Event::Exit {
+                            pid: pid.as_raw() as i32,
+                            exit_status: ExitStatus::Signalled(status as u8),
+                        })
+                        .unwrap();
+                }
+                Ok(WaitStatus::Continued(_)) | Ok(WaitStatus::Stopped(_, _)) => continue,
+                Err(Errno::ECHILD) => {
+                    // wait till a child process is spawned
+                    // unparked by the consumer thread after a Exec is received
+                    std::thread::park();
+                }
```

Comment:
If this thread is started with the first fork in this init, this branch is not needed and the Thread can just gracefully exit.

Reply:
yes that's true, but this way only spawns this thread once.

Reply:
There must be just one reaper thread. My proposal is to spawn the thread after the first fork. Keep the `JoinHandle` in an `Option`. This avoids one `waitpid` + park the thread that *will* (very very likely) happen because the thread create is probably faster than the fork.

Reply:
I'm not sure if I'm following, is the point just avoiding the first `waitpid` + `park`?